crates/re_query_cache/src/query.rs

```diff
@@ -0,0 +1,259 @@
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = &C> + '_ {
+        match self {
+            MaybeCachedComponentData::Cached(data) => itertools::Either::Left(data.iter()),
+            MaybeCachedComponentData::Raw(data) => itertools::Either::Right(data.iter()),
+        }
+    }
+
+    #[inline]
+    pub fn as_slice(&self) -> &[C] {
+        match self {
+            MaybeCachedComponentData::Cached(data) => data,
+            MaybeCachedComponentData::Raw(data) => data.as_slice(),
+        }
+    }
+}
+
+// ---
+
+/// Cached implementation of [`re_query::query_archetype`] and [`re_query::range_archetype`]
+/// (combined) for 1 point-of-view component and no optional components.
+///
+/// Alias for [`query_archetype_pov1_comp0`].
+#[inline]
+pub fn query_archetype_pov1<'a, A, R1, F>(
+    store: &'a DataStore,
+    query: &AnyQuery,
+    entity_path: &'a EntityPath,
+    f: F,
+) -> ::re_query::Result<()>
+where
+    A: Archetype + 'a,
+    R1: Component + Send + Sync + 'static,
+    F: FnMut(
+        (
+            (TimeInt, RowId),
+            MaybeCachedComponentData<'_, InstanceKey>,
+            MaybeCachedComponentData<'_, R1>,
+        ),
+    ),
+{
+    query_archetype_pov1_comp0::<A, R1, F>(store, query, entity_path, f)
+}
+
+macro_rules! impl_query_archetype {
+    (for N=$N:expr, M=$M:expr => povs=[$($pov:ident)+] comps=[$($comp:ident)*]) => { paste! {
+        #[doc = "Cached implementation of [`re_query::query_archetype`] and [`re_query::range_archetype`]"]
+        #[doc = "(combined) for `" $N "` point-of-view components and `" $M "` optional components."]
+        #[allow(non_snake_case)]
+        pub fn [<query_archetype_pov$N _comp$M>]<'a, A, $($pov,)+ $($comp,)* F>(
+            store: &'a DataStore,
+            query: &AnyQuery,
+            entity_path: &'a EntityPath,
+            mut f: F,
+        ) -> ::re_query::Result<()>
+        where
+            A: Archetype + 'a,
+            $($pov: Component + Send + Sync + 'static,)+
+            $($comp: Component + Send + Sync + 'static,)*
+            F: FnMut(
+                (
+                    (TimeInt, RowId),
+                    MaybeCachedComponentData<'_, InstanceKey>,
+                    $(MaybeCachedComponentData<'_, $pov>,)+
+                    $(MaybeCachedComponentData<'_, Option<$comp>>,)*
+                ),
+            ),
+        {
+            // NOTE: not `profile_function!` because we want them merged together.
+            re_tracing::profile_scope!(
+                "query_archetype",
+                format!("arch={} pov={} comp={}", A::name(), $N, $M)
+            );
+
+            match &query {
+                AnyQuery::LatestAt(query) => {
+                    Caches::with_latest_at::<A, _, _>(
+                        store.id().clone(),
+                        entity_path.clone(),
+                        query,
+                        |cache| {
+                            re_tracing::profile_scope!("latest_at");
```

Comment:
can we _easily_ get this out without `rustc` going nuts to fight the righwards madness?